JavaScript/chapter02/2.1 - Remove Dups/solution.js

```diff
@@ -0,0 +1,71 @@
+class Node {
+  constructor(data) {
+    this.data = data;
+    this.next = null;
+  }
+}
+
+class LinkedList {
+  constructor() {
+    this.head = null;
+  }
+
+  //Function to insert a node in the singlely linkedlist.
+  insert(data) {
+    var newNode = new Node(data);
+    if (this.head == null) {
+      this.head = newNode;
+    } else {
+      var currentNode = this.head;
+      while (currentNode.next) {
+        currentNode = currentNode.next;
+      }
+
+      currentNode.next = newNode;
+    }
+  }
+
+  //Function to print the singlely linkedlist.
+  printList() {
+    let currentNode = this.head;
+    let list = "";
+    while (currentNode) {
+      list += currentNode.data + "->";
+      currentNode = currentNode.next;
+    }
+    list += "null";
+    console.log(list);
+  }
+
+  //Function to remove dupliicate nodes from the singlely linkedlist.
+  removeDuplicates() {
```

Comment:
what is the runtime/space complexity of this function?

Reply:
The time complexity for removeDuplicates function is O(n²) and space complexity is O(1).
And time complexity for printList is O(n) and space complexity is O(1).

Reply:
is there a way to make it faster?

Reply:
Maybe by using recursion it can be made faster

Reply:
what would the space/time complexity be with recursion?

Reply:
O(n)

Reply:
can you expand on that? usually with recursive algorithms, analysis becomes complicated because of the fact that computation across method calls need to be accounted for.